File: app/test/ponte/invito_test.dart
/// Le prove dell'invito.
///
/// I vettori qui sotto sono **gli stessi** che stanno in
/// `ponte/test/invito.test.js`, ricopiati identici. Non c'e' nessun
/// compilatore che tenga insieme le due meta' di questo formato — una la
/// scrive Node, l'altra la legge Dart — e queste righe sono l'unica cosa che
/// lo faccia. Non si toccano da una parte sola.
library;

import 'package:flutter_test/flutter_test.dart';
import 'package:gdahome/ponte/indirizzo.dart';
import 'package:gdahome/ponte/invito.dart';

void main() {
  group('quello che scrive il ponte, l\'app lo rilegge', () {
    test('tutto quello che c\'e\' da dire', () {
      final letto = Invito.leggi(
        'gdahome|1|ABCD2345EFGH6789|wss://centralino.esempio.dev|'
        '192.168.1.50:8098,10.0.0.4:8098',
      );
      expect(letto.codice, 'ABCD2345EFGH6789');
      expect(
        letto.centralino,
        IndirizzoDelCentralino.leggi('wss://centralino.esempio.dev'),
      );
      expect(letto.indirizzi, [
        IndirizzoDelPonte.leggi('192.168.1.50:8098'),
        IndirizzoDelPonte.leggi('10.0.0.4:8098'),
      ]);
    });

    test('una casa senza centralino: si entra solo da dentro', () {
      final letto = Invito.leggi(
        'gdahome|1|ABCD2345EFGH6789||192.168.1.50:8098',
      );
      expect(letto.codice, 'ABCD2345EFGH6789');
      expect(letto.centralino, isNull);
      expect(letto.indirizzi, [IndirizzoDelPonte.leggi('192.168.1.50:8098')]);
    });

    test('una casa di cui non si sanno gli indirizzi', () {
      final letto = Invito.leggi(
        'gdahome|1|ABCD2345EFGH6789|wss://centralino.esempio.dev|',
      );
      expect(letto.codice, 'ABCD2345EFGH6789');
      expect(letto.centralino?.casa, 'centralino.esempio.dev');
      expect(letto.indirizzi, isEmpty);
    });

    test('il codice e basta', () {
      final letto = Invito.leggi('gdahome|1|ABCD2345EFGH6789||');
      expect(letto.codice, 'ABCD2345EFGH6789');
      expect(letto.centralino, isNull);
      expect(letto.indirizzi, isEmpty);
    });
  });

  test('perdona quello che ci mette in mezzo chi legge i quadretti', () {
    /* Un lettore di QR restituisce quello che trova, e ogni tanto ci lascia
     * attaccato un a capo. Non e' un motivo per dire di no a un codice buono. */
    final letto = Invito.leggi(
      '  gdahome|1|ABCD2345EFGH6789||192.168.1.50:8098 \n',
    );
    expect(letto.codice, 'ABCD2345EFGH6789');
    expect(letto.indirizzi, hasLength(1));

    expect(Invito.leggi('GDAHOME|1|ABCD||').codice, 'ABCD');
  });

  test('un campo in piu\' non rompe l\'app di oggi', () {
    /* E' il motivo per cui i campi stanno in coda e non in mezzo: quello che
     * arrivera' domani, un'app di oggi lo salta e va avanti. */
    final letto = Invito.leggi(
      'gdahome|1|ABCD||192.168.1.50:8098|qualcosa|che|verra\'|dopo',
    );
    expect(letto.codice, 'ABCD');
    expect(letto.indirizzi, hasLength(1));
  });

  test('un invito di domani si riconosce come tale, e lo si dice', () {
    /* La differenza che conta: «non ti capisco» manda a controllare il codice,
     * «sei vecchia» manda ad aggiornare l'app. */
    expect(
      () => Invito.leggi('gdahome|2|ABCD||'),
      throwsA(isA<InvitoTroppoNuovo>()),
    );
  });

  test('quello che non e\' un invito lo dice, invece di leggerlo a meta\'', () {
    for (final roba in [
      '',
      '   ',
      'ciao',
      'https://www.esempio.it/',
      'gdahome',
      'gdahome|1',
      'gdahome|1|',
      'gdahome|1||',
      'gdahome|x|ABCD||',
      'altracosa|1|ABCD||',
    ]) {
      expect(
        () => Invito.leggi(roba),
        throwsA(isA<InvitoIllegibile>()),
        reason: '«$roba» non e\' un invito',
      );
    }
  });

  test('un indirizzo storto si salta, e il resto dell\'invito vale', () {
    /* Una casa che dice una sciocchezza non deve far fallire l'abbinamento di
     * chi ha appena inquadrato: si tiene quello che c'e'. */
    final letto = Invito.leggi(
      'gdahome|1|ABCD|non un indirizzo|192.168.1.50:8098,,cosi\' no!,10.0.0.4:8098',
    );
    expect(letto.centralino, isNull);
    expect(letto.indirizzi, [
      IndirizzoDelPonte.leggi('192.168.1.50:8098'),
      IndirizzoDelPonte.leggi('10.0.0.4:8098'),
    ]);
  });
}

File: ponte/test/invito.test.js
/* Le prove dell'invito.
 *
 * Questo formato ha due meta' che vivono in due mondi diversi: qui lo scrive
 * il ponte, in `app/lib/ponte/invito.dart` lo legge il telefono. Non c'e'
 * nessun compilatore che le tenga insieme — l'unica cosa che le tiene insieme
 * sono **queste righe**, ricopiate identiche di la'. Se un giorno divergono,
 * una delle due prove diventa rossa prima che qualcuno inquadri un quadretto
 * che non si apre.
 */

import { test } from "node:test";
import assert from "node:assert/strict";

import { invito, leggiLInvito, InvitoIllegibile, InvitoTroppoNuovo } from "../src/invito.js";

/* Gli stessi che stanno nelle prove in Dart. Non si toccano da una parte
 * sola. */
const VETTORI = [
  {
    come: "tutto quello che c'e' da dire",
    pezzi: {
      codice: "ABCD2345EFGH6789",
      centralino: "wss://centralino.esempio.dev",
      indirizzi: ["192.168.1.50:8098", "10.0.0.4:8098"],
    },
    scritto:
      "gdahome|1|ABCD2345EFGH6789|wss://centralino.esempio.dev|192.168.1.50:8098,10.0.0.4:8098",
  },
  {
    come: "una casa senza centralino: si entra solo da dentro",
    pezzi: { codice: "ABCD2345EFGH6789", centralino: "", indirizzi: ["192.168.1.50:8098"] },
    scritto: "gdahome|1|ABCD2345EFGH6789||192.168.1.50:8098",
  },
  {
    come: "una casa di cui non si sanno gli indirizzi",
    pezzi: {
      codice: "ABCD2345EFGH6789",
      centralino: "wss://centralino.esempio.dev",
      indirizzi: [],
    },
    scritto: "gdahome|1|ABCD2345EFGH6789|wss://centralino.esempio.dev|",
  },
  {
    come: "il codice e basta",
    pezzi: { codice: "ABCD2345EFGH6789", centralino: "", indirizzi: [] },
    scritto: "gdahome|1|ABCD2345EFGH6789||",
  },
];

test("scrive quello che le prove in Dart si aspettano di leggere", () => {
  for (const uno of VETTORI) {
    assert.equal(invito(uno.pezzi), uno.scritto, uno.come);
  }
});

test("e rilegge quello che ha scritto", () => {
  for (const uno of VETTORI) {
    assert.deepEqual(leggiLInvito(uno.scritto), uno.pezzi, uno.come);
  }
});

test("perdona quello che ci mette in mezzo chi legge i quadretti", () => {
  /* Un lettore di QR restituisce quello che trova, e ogni tanto ci lascia
   * attaccato un a capo. Non e' un motivo per dire di no a un codice buono. */
  const letto = leggiLInvito("  gdahome|1|ABCD2345EFGH6789||192.168.1.50:8098 \n");
  assert.equal(letto.codice, "ABCD2345EFGH6789");
  assert.deepEqual(letto.indirizzi, ["192.168.1.50:8098"]);

  assert.equal(leggiLInvito("GDAHOME|1|ABCD||").codice, "ABCD");
});

test("un campo in piu' non rompe le app di oggi", () => {
  /* E' il motivo per cui i campi stanno in coda e non in mezzo: quello che
   * arrivera' domani, un'app di oggi lo salta e va avanti. */
  const letto = leggiLInvito("gdahome|1|ABCD||192.168.1.50:8098|qualcosa|che|verra'|dopo");
  assert.equal(letto.codice, "ABCD");
  assert.deepEqual(letto.indirizzi, ["192.168.1.50:8098"]);
});

test("un invito di domani si riconosce come tale, e lo si dice", () => {
  /* La differenza che conta: «non ti capisco» manda a controllare il codice,
   * «sei vecchio» manda ad aggiornare. Sono due strade diverse, e indovinare
   * quale sia tocca a noi, non all'utente. */
  assert.throws(() => leggiLInvito("gdahome|2|ABCD||"), InvitoTroppoNuovo);
});

test("quello che non e' un invito lo dice, invece di leggerlo a meta'", () => {
  for (const roba of [
    "",
    "   ",
    "ciao",
    "https://www.esempio.it/",
    "gdahome",
    "gdahome|1",
    "gdahome|1|",
    "gdahome|1||",
    "gdahome|x|ABCD||",
    "altracosa|1|ABCD||",
  ]) {
    assert.throws(() => leggiLInvito(roba), InvitoIllegibile, `«${roba}» non e' un invito`);
  }
});

test("un invito senza codice non si scrive nemmeno", () => {
  assert.throws(() => invito({ codice: "" }), InvitoIllegibile);
  assert.throws(() => invito({}), InvitoIllegibile);
});

test("gli spazi intorno ai pezzi si tolgono da soli", () => {
  assert.equal(
    invito({ codice: " ABCD ", centralino: " wss://c ", indirizzi: [" 10.0.0.1:8098 ", ""] }),
    "gdahome|1|ABCD|wss://c|10.0.0.1:8098",
  );
});
